### PR TITLE
Final pass and conclusion

### LINKED ISSUE #33: Guidance on using sparse AD
List decision criteria in the conclusion:

- which operators (gradient vs hessian or jacobian)
- what you want to do with your matrix (one product vs several products or inverse)
- sparsity level of your matrix (linked to partial separability). Rephrase partial separability as a way to quantify the complexity of sparsity detection, instead of something we know a priori. See #22 
- how expensive one matrix-vector product is
- how expensive sparsity detection and coloring are (even though they are amortized)
- number of colors necessary (constant or increasing?)

### LINKED ISSUE #47: Introduce or remove the word "materialize"

### LINKED ISSUE #48: Remove references to "dense AD" and "sparse AD", use AD and ASD
This also applies to `dense_backend` and `sparse_backend`.

### PR DESCRIPTION
**Pull request by gdalle**
- Write the conclusion on when to use ASD (fix #33)
- Replace dense/sparse AD with AD/ASD (fix #48)
- Remove all uses of "materialize" (fix #47)
- Add a few references, fix some bibliography entries
- Update TOC and check it
- Stylistic improvements, in particular to keep one word to a specific use
  - "Operator" mostly refers to linear operators in our blog, so for the computational graph I used "primitive functions" instead
  - Add more mentions of "structural orthogonality" instead of just "orthogonality"
- Language fixes:
  - I think one writes "using AD in reverse mode" but "using reverse-mode AD" (note the dash)

### CODE REVIEW

**Review by adrhill — CHANGES_REQUESTED** [review]
Thanks! These are great tweaks and I've added a few suggestions of mine on top.

Yesterday, we both agreed that the word "materialize" either needs to be introduced or removed, as it is quite vague.

However, I find the replacements "recover", "reconstruct", "retrieve" and "getting our hands on" even more vague.
In addition to being more vague, the first three "re-" terms imply the Jacobian was available at some point.
The fourth term "getting our hands on" is also too informal for a conference submission.

This is a very important detail to nail in our blog post, as it was the number one topic people had trouble understanding during internal presentations of this topic.

File: _posts/2025-04-28-sparse-autodiff.md [hunk 1]
```diff
@@ -926,22 +933,25 @@ As we can see on Figure 22, there are three main regimes:
...
 
 Although the specific thresholds between regimes are problem-dependent, our conclusions hold in general.
 
 ## Conclusion
 
-*TODO*
-
-When to use sparse AD? List of criteria.
+By now, the reader should have a better understanding of how sparsity can be used for efficient differentiation.
+But should it always be used? Of course not.
```
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:946]: See https://github.com/adrhill/iclr2025-sparse-autodiff/pull/51#discussion_r1853879661.

File: _posts/2025-04-28-sparse-autodiff.md [hunk 2]
```diff
@@ -194,7 +197,7 @@ the **multivariate chain rule** tells us that we obtain the Jacobian of $f$ by *
 $$ \Jfc = \Jhc \cdot \Jgc .$$
 
 Figure 1 illustrates this for $n=5$, $m=4$ and $p=3$.
-We will keep using these dimensions in following illustrations.
+We will keep using these dimensions in following illustrations, even though the real benefits of ASD only appear for large matrices.
```
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:200]:
```suggestion
We will keep using these dimensions in following illustrations, even though the real benefits of ASD only appear as the dimension grows.
```

[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:200]: let's go back to the initial formulation

File: _posts/2025-04-28-sparse-autodiff.md [hunk 3]
```diff
@@ -250,27 +252,27 @@ Our illustrations distinguish between materialized matrices and linear maps by u
...
 We visualize "matrix entries" in linear maps to build intuition.
 Even though following illustrations will sometimes put numbers onto these entries,
-linear maps are best thought of as black-box functions.
+linear maps are best thought of as black-box operators.
 </aside>
 
 ### Forward-mode AD
 
-Now that we have translated the compositional structure of our function $f$ into a compositional structure of linear maps, we can evaluate them by propagating **materialized vectors** through them.
+Now that we have translated the compositional structure of our function $f$ into a compositional structure of linear maps, we can evaluate them by propagating vectors through them, one layer at a time.
```
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:260]: yes but I want to keep the word "operator" for linear operators, as explained
```suggestion
Now that we have translated the compositional structure of our function $f$ into a compositional structure of linear maps, we can evaluate them by propagating vectors through them, one subfunction at a time.
```

File: _posts/2025-04-28-sparse-autodiff.md [hunk 4]
```diff
@@ -281,59 +283,60 @@ $$ \vvc{3} = \Dfc(\vvc{1}) = \Jfc \cdot \vvc{1} ,$$
...
 </div>
 
+This is commonly called a **vector-Jacobian product** (VJP) or **pullback**.
+Just like forward mode, reverse mode is also matrix-free: **no intermediate Jacobians are computed at any point**.
+
 ### From linear maps back to Jacobians
 
 The linear map formulation allows us to avoid intermediate Jacobian matrices in long chains of function compositions.
-But can we use this machinery to materialize the **Jacobian** of the composition $f$ itself?
+But can we use this machinery to recover the Jacobian of the composition $f$ itself?
```
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:304]:
```suggestion
But can we use this machinery to compute the Jacobian of the composition $f$ itself?
```

[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:304]:
```suggestion
But can we use this machinery to materialize the Jacobian of the composition $f$ itself?
```

File: _posts/2025-04-28-sparse-autodiff.md [hunk 5]
```diff
@@ -350,21 +353,21 @@ As shown in Figure 8, we refer to linear maps as "sparse linear maps" if they ma
...
 For now, we assume that the sparsity pattern of the Jacobian is always the same, regardless of the input, and that we know it ahead of time.
-We say that two columns or rows of the Jacobian matrix are orthogonal if, for every index, at most one of them has a nonzero coefficient.
-In other words, the vectors representing their sparsity patterns are structurally orthogonal.
-The dot product between these vectors is always zero, regardless of their values.
+We say that two columns or rows of the Jacobian matrix are **structurally orthogonal** if, for every index, at most one of them has a nonzero coefficient.
+In other words, the sparsity patterns of the columns are non-overlapping vectors,
+whose dot product is always zero regardless of their actual values.
 
-**The core idea of ASD is that we can materialize multiple orthogonal columns (or rows) in a single product evaluation.**
+The core idea of ASD is that **we can compute structurally orthogonal columns (or rows) with a single JVP (or VJP).**
```
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:368]:
```suggestion
The core idea of ASD is that **we can compute multiple structurally orthogonal columns (or rows) with a single JVP (or VJP).**
```

File: _posts/2025-04-28-sparse-autodiff.md [hunk 6]
```diff
@@ -417,22 +420,22 @@ We can then materialize multiple rows in a single VJP.
...
+### Pattern detection and coloring
 
-Unfortunately, our initial assumption had a major flaw: 
-since AD only gives us a composition of linear maps and linear maps are black-box functions,
+Unfortunately, our initial assumption had a major flaw.
+Since AD only gives us a composition of linear maps and linear maps are black-box functions,
 the structure of the Jacobian is completely unknown.
-In other words, **we cannot tell which rows and columns are orthogonal without first materializing a Jacobian matrix.**
-But if we fully materialize a Jacobian via traditional AD, ASD isn't needed at all.
+In other words, **we cannot tell which rows and columns form structurally orthogonal groups** without first getting our hands on a Jacobian matrix.
```
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:431]: I also wanted to highlight the fact that forming groups itself is not obvious. Let's keep this.

[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:431]:
```suggestion
In other words, **we cannot tell which rows and columns form structurally orthogonal groups** without first obtaining a Jacobian matrix.
```

File: _posts/2025-04-28-sparse-autodiff.md [hunk 7]
```diff
@@ -417,22 +420,22 @@ We can then materialize multiple rows in a single VJP.
...
+But if we compute this Jacobian via traditional AD, then ASD isn't necessary anymore.
 
 The solution to this problem is shown in Figure 12 (a):
-in order to find orthogonal columns (or rows), we don't need to materialize the full Jacobian.
+in order to find structurally orthogonal columns (or rows), we don't need to calculate every coefficient in the Jacobian.
```
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:435]: then the "full" Jacobian is equally confusing

File: _posts/2025-04-28-sparse-autodiff.md [hunk 8]
```diff
@@ -464,22 +467,23 @@ We now discuss the first two steps in more detail.
...
 
-Mirroring the diversity of existing approaches to AD,
+Mirroring the diversity of AD systems,
 there are also many possible approaches to sparsity pattern detection,
-each with their own advantages and tradeoffs.
-The work of Dixon <d-cite key="dixonAutomaticDifferentiationLarge1990"></d-cite> in the 1990's was the first of many papers on this subject, most of which can be classified into operator overloading or source transformation techniques.
-There are also ways to detect a sparsity pattern by probing the Jacobian with standard AD<d-cite key="griewankDetectingJacobianSparsity2002"></d-cite> but we do not linger on them here.
+each with its own advantages and tradeoffs.
+The work of Dixon <d-cite key="dixonAutomaticDifferentiationLarge1990"></d-cite> in the 1990's was among the first papers on this subject.
+After that, most the literature can be classified into operator overloading or source transformation techniques.
```
Comment (gdalle) [inline _posts/2025-04-28-sparse-autodiff.md:478]:
```suggestion
The work of Dixon <d-cite key="dixonAutomaticDifferentiationLarge1990"></d-cite> in the 1990's was among the first of many papers on this subject,
most of which can be classified into operator overloading or source transformation techniques.
```

File: _posts/2025-04-28-sparse-autodiff.md [hunk 9]
```diff
@@ -585,41 +591,42 @@ $$
...
 ### Graph formulation
 
-Luckily, this can be reformulated as a graph coloring problem, which is very well studied.
+Luckily, this grouping problem can be reformulated as graph coloring, which is very well studied.
 Let us build a graph $\mathcal{G} = (\mathcal{V}, \mathcal{E})$ with vertex set $\mathcal{V}$ and edge set $\mathcal{E}$, 
 such that each column is a vertex of the graph, and two vertices are connected if and only if their respective columns share a non-zero index.
-Put differently, an edge between vertices $j_1$ and $j_2$ means that columns $j_1$ and $j_2$ are not orthogonal.
-There are more efficient graph representations <d-cite key="gebremedhinWhatColorYour2005"></d-cite> but here we pick the simplest one.
+Put differently, an edge between vertices $j_1$ and $j_2$ means that columns $j_1$ and $j_2$ are not structurally orthogonal.
+There are more efficient graph representations <d-cite key="gebremedhinWhatColorYour2005"></d-cite> but here we picked the simplest one.
```
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:619]:
```suggestion
Note that there are more efficient graph representations, summed up in <d-cite key="gebremedhinWhatColorYour2005"></d-cite>.
```

File: _posts/2025-04-28-sparse-autodiff.md [hunk 10]
```diff
@@ -651,72 +658,74 @@ A crucial hyperparameter is the choice of ordering, for which various criteria h
 
 A more advanced coloring technique called **bicoloring** allows combining forward and reverse modes, because the recovery of the Jacobian leverages both columns (JVPs) and rows (VJPs) <d-cite key="hossainComputingSparseJacobian1998"></d-cite> <d-cite key="colemanEfficientComputationSparse1998"></d-cite>.
 
-Figure 19 shows bicoloring on a toy example in which all rows and columns are non-orthogonal.
-Materializing the Jacobian would either take $5$ JVPs in forward-mode or $4$ VJPs in reverse mode.
-Using ASD and bicoloring, we can recover the full Jacobian by computing only one JVP and one VJP.   
+Figure 19 shows bicoloring on a toy example in which no pair of columns or rows is structurally orthogonal.
+Even with ASD, the Jacobian computation would require $5$ JVPs in forward-mode or $4$ VJPs in reverse mode.
+If we use both modes simultaneously, we can recover the full Jacobian by computing only $1$ JVP and $1$ VJP.
```
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:663]: So the limit is between one and 4?
Here it's easier to see the contrast if we write everything as numbers

File: _posts/2025-04-28-sparse-autodiff.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:435]:
```suggestion
in order to find structurally orthogonal columns (or rows), we don't need to calculate the Jacobian.
```
Is this better?

File: _posts/2025-04-28-sparse-autodiff.md [hunk 11]
```diff
@@ -651,72 +658,74 @@ A crucial hyperparameter is the choice of ordering, for which various criteria h
...
-Efficient HVPs were first considered by Pearlmutter <d-cite key="pearlmutterFastExactMultiplication1994"></d-cite> and recently revisited in a 2024 ICLR blog post <d-cite key="dagreouHowComputeHessianvector2024"></d-cite>.
+Note that the gradient function is itself computed via a VJP of $f$.
+Thus, the HVP approach we described stacks a JVP on top of a VJP, which justifies the name "forward over reverse".
+In forward-over-reverse HVPs, the complexity of a single product scales roughly with the complexity of the function $f$ itself.
+This procedure was first considered by Pearlmutter <d-cite key="pearlmutterFastExactMultiplication1994"></d-cite> and recently revisited in a 2024 ICLR blog post <d-cite key="dagreouHowComputeHessianvector2024"></d-cite>.
+Note that other mode combinations are possible, like "forward over forward", which has a higher complexity but is less expensive in terms of storage.
 
 The Hessian has a **symmetric** structure (equal to its transpose), which means that matrix-vector products and vector-matrix products coincide.
+It explains why we don't need a VHP in addition to the HVP.
```
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:700]: Nope, two propositions starting with "which" in the same sentence is a lot. Sometimes you have to accept two sentences ^^

File: _posts/2025-04-28-sparse-autodiff.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:478]: This diff looks very weird in here.

File: _posts/2025-04-28-sparse-autodiff.md [hunk 12]
```diff
@@ -727,16 +736,16 @@ The closest counterpart we know is coded in C, namely the combination of ADOL-C
...
 - [ForwardDiff.jl](https://github.com/JuliaDiff/ForwardDiff.jl) <d-cite key="revelsForwardModeAutomaticDifferentiation2016"></d-cite>: Forward-mode AD and computation of JVPs.
-- [DifferentiationInterface.jl](https://github.com/JuliaDiff/DifferentiationInterface.jl) <d-cite key="dalleJuliaDiffDifferentiationInterfacejlDifferentiationInterfacev06232024"></d-cite>: High-level interface bringing all of these together.
+- [DifferentiationInterface.jl](https://github.com/JuliaDiff/DifferentiationInterface.jl) <d-cite key="dalleJuliaDiffDifferentiationInterfacejlDifferentiationInterfacev06232024"></d-cite>: High-level interface bringing all of these together, originally inspired by <d-cite key="schaferAbstractDifferentiationjlBackendAgnosticDifferentiable2022"></d-cite>.
 
 This modular pipeline comes as a replacement and extension of a previous ASD system in Julia <d-cite key="gowdaSparsityProgrammingAutomated2019"></d-cite>.
 
-As with any language, the first step is importing the dependencies.
+Like in any other language, the first step is importing the dependencies.
```
Comment (gdalle) [inline _posts/2025-04-28-sparse-autodiff.md:748]:
```suggestion
We also use a few other packages for data manipulation <d-cite key="bouchet-valatDataFramesjlFlexibleFast2023"></d-cite> and visualization <d-cite key="danischMakiejlFlexibleHighperformance2021"></d-cite>.

Like in any other language, the first step is importing the dependencies.
```

File: _posts/2025-04-28-sparse-autodiff.md [hunk 13]
```diff
@@ -727,16 +736,16 @@ The closest counterpart we know is coded in C, namely the combination of ADOL-C
 ### Necessary packages
 
 Here are the packages we will use for this demonstration.
-We also use a few other packages for visualization.
+We also use a few other packages for data manipulation <d-cite key="bouchet-valatDataFramesjlFlexibleFast2023"></d-cite> and visualization <d-cite key="danischMakiejlFlexibleHighperformance2021"></d-cite>.
```
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:739]:
```suggestion
```

File: _posts/2025-04-28-sparse-autodiff.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:478]: image

File: _posts/2025-04-28-sparse-autodiff.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:946]:
```suggestion
```

File: _posts/2025-04-28-sparse-autodiff.md [hunk 14]
```diff
@@ -926,22 +933,25 @@ As we can see on Figure 22, there are three main regimes:
...
-When to use sparse AD? List of criteria.
+By now, the reader should have a better understanding of how sparsity can be used for efficient differentiation.
+But should it always be used? Of course not.
 
-...
+Here are a list of criteria to consider when choosing between AD and ASD:
 
-Another importand aspect is the sparsity pattern of the matrix at hand, which directly influences the cost of detection, coloring and compressed differentiation.
-While it may be hard to get an exaxt estimate in realistic cases, concepts such as **partial separability** provide simple scenarios where the asymptotic complexity can be controlled explicitly <d-cite key="gowerComputingSparsityPattern2014"></d-cite>.
+- **Which derivative is needed?** In a purely gradient-based algorithm, sparsity will be useless: it can only speed up derivatives like the Jacobian and Hessian which have a matrix form.
```
[the remark this answers is not in the document]
Comment (adrhill) [reply _posts/2025-04-28-sparse-autodiff.md:949]: This is not just about NN parameters, but general ASD.
A function could very well have zero-derivatives w.r.t. some inputs, especially when using ASD with local patterns, no?

[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:949]: If you're doing gradient descent and your derivatives are zero with respect to some inputs globally, then you might as well do gradient descent on a subset of the parameters. I'm not sure forward-mode gradients are relevant at all here, especially for an ML crowd

[the remark this answers is not in the document]
Comment (gdalle) [reply _posts/2025-04-28-sparse-autodiff.md:949]: It's a tutorial. It does not have to be precise everywhere. In fact it's better to paint broad strokes than go into the smallest of details to ensure absolute and universal correctness.
